colossalai/nn/optimizer/distributed_adafactor.py

```diff
@@ -0,0 +1,336 @@
+        """
+        for group in self.param_groups:
+            # update weight & bias
+            for p in group["params"]:
+                if p.grad is None:
+                    continue
+                grad = p.grad
+                if grad.is_sparse:
+                    raise RuntimeError("Adafactor does not support sparse gradients.")
+
+                state = self.state[p]
+                grad_shape = self.grad_shape_dict[id(p)]
+                param_is_dtensor = self.param_is_dtensor_dict[id(p)]
+                if param_is_dtensor:
+                    grad_shape = self.shard_to_param.get(id(p)).shape  # tp shape (2 dim)
+                factored, use_first_moment = self.factored_dict[id(p)], self.use_first_moment_dict[id(p)]
+                shard_spec = self.shard_spec_dict[id(p)]
+                if len(state) == 0:
+                    state["step"] = 0
+                    if use_first_moment:
+                        # Exponential moving average of gradient values
+                        state["exp_avg"] = torch.zeros_like(p)
+                    if factored and param_is_dtensor:
+                        if shard_spec.sharding_sequence[0] == "R":  # Col Parallel
+                            state["exp_avg_sq_row"] = torch.zeros(
+                                grad_shape[0] // self.data_parallel_size, device=p.device, dtype=p.dtype
+                            )  # [H/dp]
+                            state["exp_avg_sq_col"] = torch.zeros(
+                                grad_shape[1], device=p.device, dtype=p.dtype
+                            )  # [W/TP]
+
+                        if shard_spec.sharding_sequence[-1] == "R":  # Row Parallel
+                            # Row indivisible shape situation
+                            if grad_shape[0] % self.data_parallel_size != 0:
+                                state["exp_avg_sq_row"] = torch.zeros(
+                                    grad_shape[0], device=p.device, dtype=p.dtype
+                                )  # [H/dp/Tp]
+                            else:
+                                state["exp_avg_sq_row"] = torch.zeros(
+                                    grad_shape[0] // self.data_parallel_size, device=p.device, dtype=p.dtype
+                                )  # [H/dp/Tp]
+
+                            state["exp_avg_sq_col"] = torch.zeros(grad_shape[1], device=p.device, dtype=p.dtype)  # [W]
+                    else:
+                        state["exp_avg_sq"] = torch.zeros_like(p)
+                    state["RMS"] = 0
+                else:
+                    if use_first_moment:
+                        state["exp_avg"] = state["exp_avg"].to(grad)
+                    if factored:
+                        state["exp_avg_sq_row"] = state["exp_avg_sq_row"].to(grad)
+                        state["exp_avg_sq_col"] = state["exp_avg_sq_col"].to(grad)
+                    else:
+                        state["exp_avg_sq"] = state["exp_avg_sq"].to(grad)
```

Comment:
DON'T MOVE STATE IN OPTIMIZER

Reply:
> DON'T MOVE STATE IN OPTIMIZER

Fixed in https://github.com/hpcaitech/ColossalAI/pull/5484/commits/c84fb52cce6f3af28bcd1b36c17276deb6d3aaa2;
Also, remove the state moving in base adafactor;
I have a question, if I do this for two state that are on the same device, does it cost additional overhead?